[WIP]model/messages/api_types: alert_words mark up

Subhasish-Behera opened the pull request:
<!-- Please see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->

**What does this PR do?**  <!-- Overall description goes here -->

<!-- If fixing a filed bug or new feature, add 'Fixes #<issue>' or 'Partial fix for #<issue>' -->
Partial fix for #588
<!-- Add a link to a discussion on chat.zulip.org, if relevant -->

**Tested?** <!-- Fine to leave some of these unchecked if this is a draft/work-in-progress -->
- [X] Manually
- [X] Existing tests (adapted, if necessary)
- [X] New tests added (for any new behavior)
- [X] Passed linting & tests (each commit)
<!-- Code must pass CI (GitHub Actions) before merging - look for the green tick! -->

<!-- See https://github.com/zulip/zulip-terminal#commit-style -->
**Commit flow** <!-- if more than one commit; add/delete/fill-in as appropriate -->
<!-- For example:
- first commit doing some thing
- maybe multiple commits doing similar things
-->
first commit:Fetches alert words and registers for any `alert_word` event and updates the alerts words.
second commit: adds transform_content_alert_words which adds `<span class=alert-word>` around alert-words

**Notes** <!-- if any; add/delete/fill-in as appropriate -->
<!-- For example:
- this doesn't include feature X (yet?)
- unsure about Y
- should this do Z?
-->
-> This PR(specifically the second commit) is based on the alert word implementation of the Zulip web client which was later adapted by other clients. 
link for the web implementation code: [link](https://github.com/zulip/zulip/blob/main/web/src/alert_words.js)
-> Other helpful documentations for contributors:
     1. [regex wrt python](https://docs.python.org/3/howto/regex.html#regex-howto)
     2. [re module](https://docs.python.org/3/library/re.html#functions)
-> Whats Done?
1. fetching and registering of alert_words
2. finding the alert_word patterns and applying the <span> 

-> What needs to be done?

1. Even though applying span logic is ready but the logic for applying markup for alert words inside of multiple elements is not ready. So basically the markup works for the alert-words without any other `tag`s indie `soup2markup`. Before that we have to decide whether we want to apply different markup when multiple tags are present for a alert_word or we have to choose the markup from any one of them.

2. Currently the markup applied for alert-words is same as `code`  tag i.e `msg_code` as the markup for different `zt_themes` are yet to be discussed.

3. Tests are not added as the non-test code is yet to reach a stable state.

neiljp reviewed (commented):
@Subhasish-Behera Thanks for looking at this other piece of the alert words puzzle :+1:

I've focused on the first commit, since we need the alert words first in order to make progress, at least in the application itself.

It'd be good to split the first commit into two parts:
- getting the initial data
- updating the data

While forgetting to include the updating wouldn't be great (and we have that with other data right now), alert words aren't often updated, so you should be able to manually test a new first commit easily - and even the highlighting if you wanted. That first commit could be merged sooner on it's own once it looks good.

Please note my point about the alert words being 'private' to the model; while other parts of the code simply access inside the model, ideally we want to use a method for this (and migrate the others too). That would sit in the new first commit, and could form part of a test that the data from the server comes out via that function. For that, I'd suggest you check out the API link I gave, as well as the fixtures in `tests/conftest.py`.

The updating part does look very simple, and in a data sense it should be, based on the event data, but note that a change in the data should affect how messages are rendered. That is, not just ones that are constructed based on the message content and alert words at the time, but they need updating too. We could ignore that initially, but we shouldn't forget it :)

As a final quick point, if you've not, I'd suggest installing gitlint as per the README. It should complain about your long commit title ;)
That section should also give you pointers on the general format.
If you find you need to say too much, or include too many files, that can indicate the commit can be split to be simpler.

Subhasish-Behera commented on zulipterminal/api_types.py:
```diff
@@ -273,6 +273,12 @@ class UpdateGlobalNotificationsEvent(TypedDict):
+class AlertWordEvent(TypedDict):
+    type: Literal["alert_words"]
+    alert_words: List[str]
+    id: int
```
Yep,` id` is not present for all the Events. I probably added it by looking at some non-Event class. So I will remove it.

Subhasish-Behera commented on zulipterminal/model.py:
```diff
@@ -136,11 +136,15 @@ def __init__(self, controller: Any) -> None:
+            "alert_words",  # added alert_words
         ]
 
+        self.alert_words = self.fetch_alert_words()
```
True. So I will change it to `self._alert_words` and add this function
```
  def get_alert_words(self):
        return self._alert_words
```
in `model.py`
and
`self.alerted_words = self. model.get_alert_words()`
in messages.py

neiljp replied:
This would be the general idea, yes. Note that if the data is not immutable, or not copied, this still allows the caller to modify the alert words (even accidentally).

Subhasish-Behera commented on zulipterminal/model.py:
```diff
@@ -813,6 +817,18 @@ def modernize_message_response(message: Message) -> Message:
 
         return message
 
+    def fetch_alert_words(self) -> List[str]:
+        """
+        Fetches alert words.
+        """
+        response = self.client.get_alert_words()
```
I think I am getting what you are saying. So I will totally remove `fetch_alert_words(self)` function and add `self.alert_words = self.initial_data['alert_words']` in the `_fetch_initial_data(self)` function in the `if not any(results.values()):` conditional.
This along with the the line `"alert_words"` inside `self.initial_data_to_fetch` will go in the first commit . 

And the line `("alert_words", self._handle_alert_words_event)` along with the `_handle_alert_words_event` function and change made to `api_types.py` will go in a second commit.

neiljp replied:
We currently have a hybrid approach there, it's true. I'm not sure there's a restriction on having the data initialized there, but for now I'd suggest putting it at an appropriate place in `__init__` after the data is fetched.

Subhasish-Behera replied:
cool, so I will make a change:
`self.alert_words = self.initial_data['alert_words']` in `__init__`

neiljp commented on zulipterminal/ui_tools/messages.py:
```diff
@@ -715,7 +721,11 @@ def main_view(self) -> List[Any]:
-            self.message["content"], self.model.server_url
+            self.message["content"],
+            self.model.server_url,
+            self.alerted_words,
+            "has_alert_word" in self.message["flags"]
```
These new parameters could be combined into one. If the message doesn't have alert words flagged, then the function doesn't need to know/process them.

neiljp commented on zulipterminal/ui_tools/messages.py:
```diff
@@ -796,14 +806,58 @@ def update_message_author_status(self) -> bool:
 
         return author_is_present
 
+    def transform_content_alert_words(content: str, alerted_list: List[str]) -> Any:
```
If this is translated from a server implementation, it would be useful to add a comment to note that, and in the commit text too.

That could refer to the file as it is now, but also a commit ref, so that if code changes later then it is easier to track what has changed in the original code.

Having this function in a commit on it's own first, with detailed tests to cover various cases, would be good before we hook it into the processing in a subsequent commit.

neiljp reviewed (commented):
@Subhasish-Behera Let's still address the first commit first, but I just took a look at the second commit as it stands.

I can see how this mirrors web-app code, but we should test it ourselves too, to ensure that it works how we expect in this translated form.

If we can demonstrate this manually and with tests, this should be a good v1 implementation of this feature.

Beyond that, if we have more tests, and can compare what the web app highlights (and tests), we should be able to determine whether we can do the same using this current approach. This will affect if we need to think differently to cover more cases.

We may want to consider an alternative way of formatting this kind of text, and maybe search text too, if it crosses tag/style boundaries.

I expect the web app applies tags dynamically since it uses CSS to style the text, and I'm guessing the spans can overlap? That's complex with urwid - we process the HTML into urwid-styled Text instead.

If so, we might end up considering processing the content into text only, finding matches, then mapping those into adjacent pieces in the HTML or urwid-processed text.

However, let's focus on testing this v1 first :)
(and the v0, ie. the data handling, before that!)

Commit:
messages: Added a method transform_content_alert_words.
stores alert_words using get_alert_words.Two new parameters
for tranform_content- alerted_list and alert_word_present.
if the second argument is true then a call is made to
transform_content_alert_words. There firstly all the alert_words
are converted into a string consisting of 3 parts clean,
befor_punctuation and after_puntuation.Secondly, these modified alerted words
are compiled to be searched in the content and added with the span
using replace_callback.

Subhasish-Behera commented:
@zulipbot add "PR needs review"

neiljp commented on zulipterminal/model.py:
```diff
@@ -182,6 +183,8 @@ def __init__(self, controller: Any) -> None:
 
         self._subscribe_to_streams(self.initial_data["subscriptions"])
 
+        self._alert_words = self.initial_data.get("alert_words", None)
```
While using `get` is sort of helpful here, in case it's not in the data, this would cause issues later if it were to be None. We don't validate structure but rely on the expected data being present, for the most part.

In this case, mypy doesn't complain since initial_data is a `Dict[str, Any]`, and it doesn't care what happens with the `Any` 👎.

This means we need to give the type for `self._alert_words`, and only from that point onwards is it guaranteed to be non-None, so a `.get` here makes little sense, unless we're going to do a separate assert before this - or simply let a KeyError drop out in case of an error.

Subhasish-Behera replied:
Upon looking at the `self._alert_words = self.initial_data.get("alert_words", None)` part, I saw that when the user has no alert_words, it returns an empty list(I was probably thinking `None` earlier,not sure). So changed it to `self._alert_words: List[str] = self.initial_data["alert_words"]`.

neiljp commented on zulipterminal/model.py:
```diff
@@ -339,6 +342,9 @@ def set_search_narrow(self, search_query: str) -> None:
         self.narrow.append(["search", search_query])
 
+    def get_alert_words(self) -> List[str]:
+        return self._alert_words
```
We need to return a copy here, to ensure the data isn't modified by a caller of this function.

neiljp commented on zulipterminal/model.py:
```diff
@@ -1567,6 +1568,13 @@ def _handle_message_event(self, event: Event) -> None:
 
             self.controller.update_screen()
 
+    def _handle_alert_words_event(self, event: Event) -> None:
```
Is there a better location for this function to be added? :)

I know model.py is long, but the function above uses the function below, it seems.

Subhasish-Behera replied:
Moved the position in between `_handle_subscription_event` and `_handle_typing_event`

neiljp commented on tests/model/test_model.py:
```diff
@@ -2624,6 +2624,51 @@ def test__handle_reaction_event_for_msg_in_index(
+            "Add and delete alert words",
+        ],
+    )
+    def test__handle_alert_words_event(
```
I would expect to see this test in the same commit as the new `_handle_*_event` function.

You might like to split the function+test, then connect it up with the event-handler system.

Subhasish-Behera replied:
I have squashed the 2nd and 3rd commit into one.

neiljp commented on tests/model/test_model.py:
```diff
@@ -2624,6 +2624,51 @@ def test__handle_reaction_event_for_msg_in_index(
+                ["word1", "word3"],
+            ),
+        ],
+        ids=[
+            "Add multiple alert words",
+            "Empty alert words",
+            "Add single new alert words",
+            "Delete an alert word",
+            "Add and delete alert words",
+        ],
```
Can you please place these inline using `case`?

neiljp commented on tests/model/test_model.py:
```diff
@@ -2624,6 +2624,51 @@ def test__handle_reaction_event_for_msg_in_index(
+    ):
+        model._handle_alert_words_event(event)
+
+        assert model._alert_words != initial_alerted_words
```
Is this always true?

Subhasish-Behera replied:
Yep, this is always true. But it is more context-dependent. Basically, currently, this test is designed to test the most minimal of changes in the `active_conversation_info` i.e change(`start/stop`) of a single user. The `inital_alerted_words` means just before that change and not at the start of the whole typing event series.
So it is true because of the values that are passed in `mark.parametrize` i.e it is an approach followed in the test.

neiljp replied:
If that is true, then it is good to add a suffix to the test function name to indicate that.

Subhasish-Behera replied:
I think the name `test__handle_alert_words_event_minimal_change` would do the job.

neiljp replied:
Hmm. Do we get an event if the changes aren't made?

I'd suggest a double underscore from the function name to any 'meaning'.

Subhasish-Behera replied:
I believe by changes you mean no typing in the writebox(of any client).
That would be a event with `op`=`stop`. So yes.

Subhasish-Behera replied:
`test__handle_alert_words_event_minimal_change` sounds bit subjective,so i will be changing it to `test__handle_alert_words_event__single_user`. Though it's natural as a single event can't inform about more than 1 user in this case.

neiljp reviewed (commented):
@Subhasish-Behera I focused on the first 3 commits, since I think that was the recent focus?

The current overall PR fails when I run it right now, at line 843 of messages.py, compiling a regex (multiple repeat at position 27).

Otherwise, improving your commit text would improve the speed of merging, and the first three seem close, if needing some adjustments.

Commit:
model: Add alert_words to initial_data_to_fetch.Tests adapted.
The alert words are stored in _alert_words and it's copy can
be accessed using get_alert_words.

neiljp commented on zulipterminal/model.py:
```diff
@@ -334,6 +337,9 @@ def set_search_narrow(self, search_query: str) -> None:
         self.unset_search_narrow()
         self.narrow.append(["search", search_query])
 
+    def get_alert_words(self) -> List[str]:
+        return self._alert_words.copy()
+
```
I didn't comment on this before, but this function placement in the file seems strange too - between two related functions?

Subhasish-Behera replied:
will fix it asap

neiljp replied:
Note that strictly this *could* go into a separate commit with a test to eg. ensure that one can't modify the data externally, but this is fine.

neiljp commented on tests/model/test_model.py:
```diff
@@ -2623,6 +2624,49 @@ def test__handle_reaction_event_for_msg_in_index(
+                {"type": "alert_words", "alert_words": ["word1", "word2", "word3"]},
+                ["word1"],
+                ["word1", "word2", "word3"],
+                id="Add multiple alert words",
```
minor, but to keep consistent: we tend to use underlines instead of spaces, in case of multiple parametrize.

Commit:
model/api_types: Update after alert_words event.Test adapted/added. "alert_words" added to event actions and corresponding handler is _handle_alert_words_event which updates _alert_words. Added AlertWordEvent in api_types.py Tests are added for _handle_alert_words_event.

theViz343 commented on zulipterminal/ui_tools/messages.py:
```diff
@@ -798,14 +808,59 @@ def update_message_author_status(self) -> bool:
+        for word in alerted_list:
+            clean = "".join(alert_regex_replacements.get(c, c) for c in word)
+            regex = f"({before_punctuation})({clean})({after_punctuation})"
+            regex1 = re.compile(regex, re.IGNORECASE)
+            content = re.sub(regex1, replace_callback, content)
```
The method `re.sub` has a parameter called `flags` in which you could pass the `IGNORECASE` flag, instead of defining a new regex. Ex: `re.sub(.., flags=re.IGNORECASE)`

theViz343 reviewed (commented):
@Subhasish-Behera Thanks for working on this :+1:! This seems to work well. There are a few cases I found during testing where it seems to be a bit unpredictable. 

- `\something`. Prefixing with a backslash seems to not get picked up by the regex if the alert word is `\something`. If the alert word is just `something`, it gets picked up when it shouldn't (according to the behavior on the webapp).
- `wqdnasj\` Suffixing with a backslash when the alert word is also `wqdnasj\` seems to crash ZT with the following error - `re.error: missing ), unterminated subpattern at position 23`. 

Apart from this, I've left some suggestions for minor code changes in the comments.

theViz343 commented on zulipterminal/ui_tools/messages.py:
```diff
@@ -798,14 +808,59 @@ def update_message_author_status(self) -> bool:
+        clean: str
+
+        def replace_callback(match: Match[Union[str, str, str]]) -> str:
+            before = match.group(1)
+            word = match.group(2)
+            after = match.group(3)
```
Minor: You could use `match.group(1,2,3)` here, which returns a tuple consisting of the three values.

Subhasish-Behera replied:
The only use these 3 have is in here:
```
 if in_tag:
                return f"{before}{word}{after}"
            return f"{before}<span class='alert-word'>{word}</span>{after}"
```
 I could have directly used `match. group(n)` but didn't. I think separating them into 3 variables gives better readability
 By following this suggestion of `match.group(1,2,3)`, I need it to be stored in some var and break it using the index to use it in the above code which will not clarify any further of what is being done. 
So why group it when it can be accessed individually without creating any `tuple`

neiljp replied:
If I follow the above correctly, would it help if you used tuple unpacking?
```python3
one, two, three = function_returning_tuple_of three_values()

Subhasish-Behera replied:
Yep! Multiple assignments in one line can do the same thing.
```
before,word,after = match.group(1,2,3)
```

neiljp commented:
@Subhasish-Behera Is this awaiting an update from you, or have you addressed the feedback?
Just checking based on my earlier feedback and that from Vishwesh :)
(If so, you can add a 👍 emoji or a comment to specific comments to indicate this)

Subhasish-Behera commented on zulipterminal/ui_tools/messages.py:
```diff
@@ -798,14 +808,59 @@ def update_message_author_status(self) -> bool:
+
+        after_punctuation = r"(?=\s|$|<|[\\)\"?!:.,';\]!])"
+
+        clean: str
```
Yeah! `clean` here is unrequired.

Subhasish-Behera committed:
commit 1
